skill/wallpaper.py

```diff
@@ -0,0 +1,104 @@
+    """
+
+    def __init__(self, skill_root_dir: str, skill_data_dir: str):
+        self.skill_directory = Path(skill_root_dir).joinpath("ui/wallpapers")
+        self.skill_data_directory = Path(skill_data_dir).joinpath("wallpapers")
+        self._ensure_user_directory_exists()
+        self.default = self._get_file_path(DEFAULT_WALLPAPER)
+        self.selected = self.default
+        self.collection = []
+        self.collect()
+
+    def _ensure_user_directory_exists(self):
+        """Ensures the directory for user-defined wallpapers exists."""
+        if not self.skill_data_directory.exists():
+            self.skill_data_directory.mkdir()
+
+    def collect(self):
+        """Builds a list of wallpapers provided by the skill and added by the user."""
+        for wallpaper_path in self.skill_directory.iterdir():
+            self.collection.append(wallpaper_path)
+        for wallpaper_path in self.skill_data_directory.iterdir():
+            self.collection.append(wallpaper_path)
+
+    def change(self, file_name: str):
+        """Change the wallpaper to use the image in a specified file.
+
+        Args:
+            file_name: the name of the wallpaper file
+        """
+        file_path = self._get_file_path(file_name)
+        if file_path in self.collection:
+            self.selected = file_path
+        else:
+            self.selected = self.default
+
+    def next(self):
+        """Selects the next wallpaper in the collection for display.
+
+        Starts over the beginning of the list after the last in the collection.  If
+        the "selected" attribute contains a value not in the collection, the default
+        is selected.
+        """
+        if self.selected in self.collection:
+            if self.selected == self.collection[-1]:
+                self.selected = self.collection[0]
+            else:
+                index_of_selected = self.collection.index(self.selected)
+                self.selected = self.collection[index_of_selected + 1]
+        else:
+            self.selected = self.default
+
+    def add(self, wallpaper_url: str):
+        """Adds a new wallpaper to the collection and selects it for display.
+
+        Args:
+            wallpaper_url: the url to download the wallpaper image from
+        """
+        response = requests.get(wallpaper_url)
+        file_path = self.skill_data_directory.joinpath("wallpaper.jpg")
+        with open(file_path, "wb") as wallpaper_file:
+            wallpaper_file.write(response.content)
+        self.collect()
+        self.selected = file_path
+
+    def _get_file_path(self, file_name: str):
+        """Determines if a wallpaper file name is from the user or skill directory."""
+        skill_path = self.skill_directory.joinpath(file_name)
+        user_path = self.skill_data_directory.joinpath(file_name)
+        if skill_path.exists():
+            file_path = skill_path
+        else:
+            file_path = user_path
```

Comment:
Probably doesn't matter since a non-existent file path shouldn't be in the collection, but my first reaction to this was that we aren't testing if the user_path version exists.

Reply:
good catch.  fixed.